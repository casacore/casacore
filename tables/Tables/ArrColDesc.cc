//# ArrColDesc.cc: Templated class to describe columns of arrays in tables
//# Copyright (C) 1994,1995,1996,1997,1999,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: casa-feedback@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA

#include <casacore/tables/Tables/ArrColDesc.h>
#include <casacore/tables/Tables/ArrColData.h>
#include <casacore/casa/Arrays/IPosition.h>
#include <casacore/casa/IO/AipsIO.h>
#include <casacore/casa/iostream.h>


namespace casacore { //# NAMESPACE CASACORE - BEGIN

ArrayColumnDescBase::ArrayColumnDescBase (const String& name,
                                          const String& comment,
                                          const String& dataManagerType,
                                          const String& dataManagerGroup,
                                          DataType dt, const String& dataTypeId,
                                          Int options,
                                          uInt ndim, const IPosition& shape)
  : BaseColumnDesc (name, comment, dataManagerType, dataManagerGroup,
                    dt, dataTypeId, options, ndim, shape,
                    False, True, False)
{
    if (nrdim_p <= 0) {
	nrdim_p = -1;
    }
}

ArrayColumnDescBase::ArrayColumnDescBase (const ArrayColumnDescBase& that)
: BaseColumnDesc (that)
{}

ArrayColumnDescBase::~ArrayColumnDescBase()
{}

ArrayColumnDescBase& ArrayColumnDescBase::operator=
                                       (const ArrayColumnDescBase& that)
{
    BaseColumnDesc::operator= (that);
    return *this;
}

//# Return the class name.
String ArrayColumnDescBase::className() const
    { return "ArrayColumnDesc<" + dataTypeId(); }

//# Put the object.
//# The data is read by the ctor taking AipsIO.
//# It was felt that putstart takes too much space, so therefore
//# the version is put "manually".
void ArrayColumnDescBase::putDesc (AipsIO& ios) const
{
    ios << (uInt)1;                  // class version 1
    // Formerly a switch was written to determine if a default existed.
    // This switch was always false.
    // Keep writing this switch (which is not used anymore).
    ios << False;
}

void ArrayColumnDescBase::getDesc (AipsIO& ios)
{
    uInt version;
    ios >> version;
    // Formerly a switch was written to determine if a default existed.
    // This switch was always false.
    // So read it in and do not do anything with it.
    Bool sw;
    ios >> sw;
}

//# Show the column.
void ArrayColumnDescBase::show (ostream& os) const
{
    os << "   Name=" << name();
    os << "   DataType=" << dataType();
    if (dataType() == TpOther) {
	os << ", " << dataTypeId();
    }
    if (maxLength() > 0) {
	os << "   MaxLength=" << maxLength();
    }
    os << "   Nrdim=" << ndim();
    os << "   Shape=" << shape();
    os << endl;
    os << "   DataManager=" << dataManagerType();
    os << "/" << dataManagerGroup();
    os << endl;
    os << "   Comment = " << comment() << endl;
}


//# Create a column object from the description.
PlainColumn* ArrayColumnDescBase::makeColumn (ColumnSet* csp) const
{
    PlainColumn* bcp = new ArrayColumnData (this, csp);
    return bcp;
}


} //# NAMESPACE CASACORE - END
