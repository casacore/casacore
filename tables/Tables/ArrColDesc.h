//# ArrColDesc.h: Templated class to describe columns of arrays in tables
//# Copyright (C) 1994,1995,1996,1997,1999,2000
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef TABLES_ARRCOLDESC_H
#define TABLES_ARRCOLDESC_H

//# Includes
#include <casa/aips.h>
#include <tables/Tables/BaseColDesc.h>
#include <casa/Containers/SimOrdMap.h>
#include <casa/Arrays/IPosition.h>


namespace casa { //# NAMESPACE CASA - BEGIN

//# Forward Declarations
class PlainColumn;
class ColumnSet;
template<class T> class Array;

// <summary>
// Templated class for description of table array columns
// </summary>

// <use visibility=export>

// <reviewed reviewer="Gareth Hunt" date="94Nov17" tests="">
// </reviewed>

// <prerequisite>
//   <li> BaseColumnDesc (and its prerequisites)
//   <li> TableDesc
// </prerequisite>

// <etymology>
// This class builds descriptions of table columns where each cell (which
// may also be called a row) will hold an array.
// </etymology>

// <synopsis> 
// ArrayColumnDesc is a templated class for defining a table column
// containing arrays.
//
// The table values are handled by a data manager. This can be
// a storage manager to store the values in a file or it can be
// a virtual column engine to calculate them on-the-fly.
// Only the basic data types are allowed when storing in a file. These are:
//  Bool, uChar, Short, uShort, Int, uInt, float, double,
//  Complex, DComplex and String.
//
// At table creation time (when a table gets created from a table
// description), each column needs to be bound to a data manager.
// If not done explicitly, the table system will bind a column to the
// default manager defined in the column description.
//
// An array column description consists of the following attributes:
// <ul>
//  <li> Name, which has to be unique and must also be different
//         from possible table keyword names.
//  <li> Data type, which is determined by the template parameter
//         (e.g. ArrayColumnDesc<Int>).
//  <li> A data type id, which tells the unique name of non-standard
//         data types (i.e. for data type == TpOther).
//  <li> Comment, which defaults to the empty string.
//         This serves purely as an informational string for the user.
//  <li> Dimensionality. If given, all arrays in the column need
//         to have that dimensionality.
//  <li> Shape. If given, all arrays in the column need to have
//         that shape.
//  <li> Default data manager, which will be used if a column
//         for a newly created table is not explicitly bound to a
//         datamanager.
//  <li> Data manager group, which serves 2 purposes.
//         Firstly it can be used in class SetupNewTable to bind a group
//         of columns.
//         Secondly, when the default data managers are used, it
//         allows, for example, to have 2 AipsIO storage managers.
//         One for one group of columns and one for another group of columns.
//  <li> Options. These are defined in ColumnDesc.h and can be combined
//       by logically or-ing them.
//       <ol>
//        <li>
//         ColumnDesc::FixedShape says that the arrays in all cells
//         of a column have the same shape. This shape must be defined
//         before a table is created. It does not tell if
//         the array is direct or indirect.
//         A FixedShape array is defined in every cell, while for
//         non-FixedShape arrays a cell can be empty.
//        <li>
//         ColumnDesc::Direct determines if an array is directly
//         stored in the table or if it is stored indirectly in a separate
//         file. Direct arrays enforce the FixedShape option.
//         Usually indirect arrays are only read in on command, while
//         direct arrays are held in memory. So the size of the
//         arrays is an important factor.
//       </ol>
//  <li> Default keyword set, which defaults to an empty set.
//         When a table column gets created from the description, it gets
//         a copy of this keyword set as its initial keyword set.
// </ul>
//
// There are several constructors, which allow the definition of most
// of the above mentioned attributes. Others, like the default keyword
// set, have to be defined explicitly.
//
// This class is derived from BaseColumnDesc, thus the functions
// in there also apply to this class.
//
// Once a column description is set up satisfactorily, it must be added
// to a table description before it can be used by the table system.
// </synopsis>

// <example>
// <srcblock>
//     TableDesc tabDesc("tTableDesc", "1", TableDesc::New);
//
//     // Now define array columns.
//     // This one is indirect and has no dimensionality mentioned yet.
//     // Define the keyword UNIT in it.
//     ArrayColumnDesc<Complex> arr1Column("Arr1", "comment for Arr1");
//     arr1Column.rwKeywordSet().define ("UNIT", "Jy");
//     tabDesc.addColumn (arr1Column);
//
//     // This one is indirect and has 3-dim arrays.
//     tabDesc.addColumn (ArrayColumnDesc<Int>("Arr2",
//                                             "comment for Arr2",
//                                             3));
//     // This one is direct and has 2-dim arrays with axis lengths 4 and 7.
//     tabDesc.addColumn (ArrayColumnDesc<uInt>("Arr3",
//                                              "comment for Arr1",
//        				        IPosition(2,4,7),
// 					        ColumnDesc::Direct));
// </srcblock>
// </example>

// <motivation>
// Several column description classes are needed to allow the user
// to define attributes which are special for each column type.
// For scalars the special attribute is the default value.
// They all have to be templated to support arbitrary data types.
// </motivation>

// <templating arg=T>
//  <li> Default constructor
//  <li> Copy constructor
//  <li> Assignment operator
//  <li> <src>static String dataTypeId();  // (not needed for builtin types)</src>
//       This should return the unique "name" of the class.
// </templating>

//# <todo asof="$DATE:$">
//# A List of bugs, limitations, extensions or planned refinements.
//# </todo>

template<class T>
class ArrayColumnDesc : public BaseColumnDesc
{
friend class ColumnDesc;

public:
    // Construct the column with the given name and dimensionality.
    // The data manager type defaults to the StandardStman storage manager.
    // The data manager group defaults to the data manager type.
    // Ndim <=0 means that the number of dimensions is free and will
    // be defined when creating the table (rows). Ndim>0 means that
    // the arrays in this column must have the given dimensionality.
    // The possible options are defined in ColumnDesc.h.
    explicit ArrayColumnDesc (const String& name, Int ndim = -1,
			      int options = 0);

    // Construct the column with the given name, dimensionality, and comment.
    // The data manager type defaults to the StandardStman storage manager.
    // The data manager group defaults to the data manager type.
    // Ndim <=0 means that the number of dimensions is free and will
    // be defined when creating the table (rows). Ndim>0 means that
    // the arrays in this column must have the given dimensionality.
    // The possible options are defined in ColumnDesc.h.
    ArrayColumnDesc (const String& name, const String& comment,
		     Int ndim = -1, int options = 0);

    // Construct the column with the given name, dimensionality, comment,
    // and default data manager type and group.
    // A blank data manager group defaults to the data manager type.
    // Ndim <=0 means that the number of dimensions is free and will
    // be defined when creating the table (rows). Ndim>0 means that
    // the arrays in this column must have the given dimensionality.
    // The possible options are defined in ColumnDesc.h.
    ArrayColumnDesc (const String& name, const String& comment,
		     const String& dataManName, const String& dataManGroup,
		     Int ndim = -1, int options = 0);

    // Construct the column with the given name and shape.
    // The data manager type defaults to the StandardStman storage manager.
    // The data manager group defaults to the data manager type.
    // The possible options are defined in ColumnDesc.h.
    // This constructor can only be used for FixedShape arrays, because the
    // shape of other arrays can only be set per row.
    ArrayColumnDesc (const String& name,
		     const IPosition& shape, int options = 0);

    // Construct the column with the given name, shape, and comment.
    // The data manager type defaults to the StandardStman storage manager.
    // The data manager group defaults to the data manager type.
    // The possible options are defined in ColumnDesc.h.
    // This constructor can only be used for FixedShape arrays, because the
    // shape of other arrays can only be set per row.
    ArrayColumnDesc (const String& name, const String& comment,
		     const IPosition& shape, int options = 0);

    // Construct the column with the given name, shape, comment,
    // and default data manager type and group.
    // A blank data manager group defaults to the data manager type.
    // The possible options are defined in ColumnDesc.h.
    // This constructor can only be used for FixedShape arrays, because the
    // shape of other arrays can only be set per row.
    // If both ndim and shape are given as > 0, ndim should match the length
    // of shape.
    ArrayColumnDesc (const String& name, const String& comment,
		     const String& dataManName, const String& dataManGroup,
		     const IPosition& shape, int options = 0, int ndim=-1);

    // Copy constructor (copy semantics);
    ArrayColumnDesc (const ArrayColumnDesc<T>&);

    ~ArrayColumnDesc();

    // Assignment (copy semantics);
    ArrayColumnDesc<T>& operator= (const ArrayColumnDesc<T>&);

    // Clone this column description to another.
    BaseColumnDesc* clone() const;

    // Get the name of this class. It is used by the registration process.
    // The template argument gets part of the name.
    String className() const;

    // Create a Column object out of this.
    // This is used by class ColumnSet to construct a table column object.
    virtual PlainColumn* makeColumn (ColumnSet*) const;

    // Show the column.
    void show (ostream& os) const;

public:
    // Construct the object to register its class.
    ArrayColumnDesc
      (SimpleOrderedMap<String, BaseColumnDesc* (*)(const String&)>&);

protected:
    // Create the object from AipsIO (this function is registered).
    static BaseColumnDesc* makeDesc(const String& name);

    // Put the object.
    virtual void putDesc (AipsIO&) const;

    // Get the object.
    virtual void getDesc (AipsIO&);
};



} //# NAMESPACE CASA - END

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <tables/Tables/ArrColDesc.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
