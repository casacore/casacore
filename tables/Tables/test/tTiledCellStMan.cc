//# tTiledCellStMan.cc: Test program for the TiledCellStMan classes
//# Copyright (C) 1994,1995,1996,1999,2000,2001,2003
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <tables/Tables/TableDesc.h>
#include <tables/Tables/SetupNewTab.h>
#include <tables/Tables/Table.h>
#include <tables/Tables/ScaColDesc.h>
#include <tables/Tables/ArrColDesc.h>
#include <tables/Tables/ScalarColumn.h>
#include <tables/Tables/ArrayColumn.h>
#include <tables/Tables/TiledCellStMan.h>
#include <casa/Arrays/Vector.h>
#include <casa/Arrays/Matrix.h>
#include <casa/Arrays/ArrayMath.h>
#include <casa/Arrays/ArrayLogical.h>
#include <casa/Arrays/ArrayUtil.h>
#include <casa/Exceptions/Error.h>
#include <casa/iostream.h>

#include <casa/namespace.h>
// <summary>
// Test program for the TiledCellStMan class.
// </summary>

// This program tests the class TiledCellStMan and related classes.
// The results are written to stdout. The script executing this program,
// compares the results with the reference output file.

void writeFixed(const TSMOption&);
void readTable(const TSMOption&);
void writeVar(const TSMOption&);
void writeFixVar(const TSMOption&);
void writeNoHyper(const TSMOption&);

int main () {
    try {
        writeFixed(TSMOption::Cache);
	readTable(TSMOption::Cache);
	writeVar(TSMOption::Buffer);
	readTable(TSMOption::MMap);
	writeFixVar(TSMOption::Cache);
	readTable(TSMOption::Buffer);
	writeNoHyper(TSMOption::MMap);
	readTable(TSMOption::Cache);
    } catch (AipsError x) {
	cout << "Caught an exception: " << x.getMesg() << endl;
	return 1;
    } 
    return 0;                           // exit with success status
}

// First build a description.
void writeFixed(const TSMOption& tsmOpt)
{
    // Build the table description.
    TableDesc td ("", "1", TableDesc::Scratch);
    td.addColumn (ArrayColumnDesc<float>  ("Pol", IPosition(1,16),
					   ColumnDesc::FixedShape));
    td.addColumn (ArrayColumnDesc<float>  ("Freq", 1, ColumnDesc::FixedShape));
    td.addColumn (ArrayColumnDesc<float>  ("Data", 2, ColumnDesc::FixedShape));
    td.addColumn (ArrayColumnDesc<float>  ("Weight", IPosition(2,16,25),
					   ColumnDesc::FixedShape));
    td.defineHypercolumn ("TSMExample",
			  2,
			  stringToVector ("Data,Weight"),
			  stringToVector ("Pol,Freq"));
    
    // Now create a new table from the description.
    SetupNewTable newtab("tTiledCellStMan_tmp.data", td, Table::New);
    // Create a storage manager for it.
    TiledCellStMan sm1 ("TSMExample", IPosition(2,5,6));
    newtab.setShapeColumn ("Freq", IPosition(1,25));
    newtab.setShapeColumn ("Data", IPosition(2,16,25));
    newtab.bindAll (sm1);
    Table table(newtab, 0, False, Table::LittleEndian, tsmOpt);

    Vector<float> freqValues(25);
    Vector<float> polValues(16);
    indgen (freqValues, float(200));
    indgen (polValues, float(300));
    ArrayColumn<float> freq (table, "Freq");
    ArrayColumn<float> pol (table, "Pol");
    ArrayColumn<float> data (table, "Data");
    ArrayColumn<float> weight (table, "Weight");
    Matrix<float> array(IPosition(2,16,25));
    Matrix<float> result(IPosition(2,16,25));
    uInt i;
    indgen (array);
    for (i=0; i<101; i++) {
	table.addRow();
	data.put (i, array);
	weight.put (i, array+float(100));
	freq.put (i, freqValues);
	pol.put (i, polValues);
	array += float(200);
	freqValues += float(200);
	polValues += float(200);
    }
    indgen (array);
    indgen (freqValues, float(200));
    indgen (polValues, float(300));
    for (i=0; i<table.nrow(); i++) {
	data.get (i, result);
	if (! allEQ (array, result)) {
	    cout << "mismatch in data row " << i << endl;
	}
	weight.get (i, result);
	if (! allEQ (array + float(100), result)) {
	    cout << "mismatch in weight row " << i << endl;
	}
	if (! allEQ (freq(i), freqValues)) {
	    cout << "mismatch in freq row " << i << endl;
	}
	if (! allEQ (pol(i), polValues)) {
	    cout << "mismatch in pol row " << i << endl;
	}
	array += float(200);
	freqValues += float(200);
	polValues += float(200);
    }
}

void readTable(const TSMOption& tsmOpt)
{
    Table table("tTiledCellStMan_tmp.data", Table::Old, tsmOpt);
    ArrayColumn<float> freq (table, "Freq");
    ArrayColumn<float> pol (table, "Pol");
    ArrayColumn<float> data (table, "Data");
    ArrayColumn<float> weight (table, "Weight");
    Vector<float> freqValues(25);
    Vector<float> polValues(16);
    indgen (freqValues, float(200));
    indgen (polValues, float(300));
    Matrix<float> array(IPosition(2,16,25));
    Matrix<float> result(IPosition(2,16,25));
    uInt i;
    indgen (array);
    for (i=0; i<table.nrow(); i++) {
	data.get (i, result);
	if (! allEQ (array, result)) {
	    cout << "mismatch in data row " << i << endl;
	}
	weight.get (i, result);
	if (! allEQ (array + float(100), result)) {
	    cout << "mismatch in weight row " << i << endl;
	}
	if (! allEQ (freq(i), freqValues)) {
	    cout << "mismatch in freq row " << i << endl;
	}
	if (! allEQ (pol(i), polValues)) {
	    cout << "mismatch in pol row " << i << endl;
	}
	array += float(200);
	freqValues += float(200);
	polValues += float(200);
    }
}

void writeVar(const TSMOption& tsmOpt)
{
    // Build the table description.
    TableDesc td ("", "1", TableDesc::Scratch);
    td.addColumn (ArrayColumnDesc<float>  ("Pol", 1));
    td.addColumn (ArrayColumnDesc<float>  ("Freq", 1));
    td.addColumn (ArrayColumnDesc<float>  ("Data", 2));
    td.addColumn (ArrayColumnDesc<float>  ("Weight", 2));
    td.defineHypercolumn ("TSMExample",
			  2,
			  stringToVector ("Data,Weight"),
			  stringToVector ("Pol,Freq"));
    
    // Now create a new table from the description.
    SetupNewTable newtab("tTiledCellStMan_tmp.data", td, Table::New);
    // Create a storage manager for it.
    TiledCellStMan sm1 ("TSMExample", IPosition(2,5,6));
    newtab.bindAll (sm1);
    Table table(newtab, 0, False, Table::BigEndian, tsmOpt);

    Vector<float> freqValues(25);
    Vector<float> polValues(16);
    indgen (freqValues, float(200));
    indgen (polValues, float(300));
    ArrayColumn<float> freq (table, "Freq");
    ArrayColumn<float> pol (table, "Pol");
    ArrayColumn<float> data (table, "Data");
    ArrayColumn<float> weight (table, "Weight");
    Matrix<float> array(IPosition(2,16,25));
    Matrix<float> result(IPosition(2,16,25));
    uInt i;
    indgen (array);
    for (i=0; i<5; i++) {
	table.addRow();
	cout << " pol.isDefined=" << pol.isDefined(i) << endl;
	pol.setShape (i, IPosition(1,16), IPosition(1,1));
	cout << " pol.isDefined=" << pol.isDefined(i) << endl;
	cout << "data.isDefined=" << data.isDefined(i) << endl;
	data.setShape (i, IPosition(2,16,25), IPosition(2,12,10));
	cout << "weig.isDefined=" << weight.isDefined(i) << endl;
	cout << "freq.isDefined=" << freq.isDefined(i) << endl;
	cout << pol.shape(i) << freq.shape(i) << data.shape(i)
	     << weight.shape(i) << endl;
	data.put (i, array);
	weight.put (i, array+float(100));
	freq.put (i, freqValues);
	pol.put (i, polValues);
	array += float(200);
	freqValues += float(200);
	polValues += float(200);
    }
}

void writeFixVar(const TSMOption& tsmOpt)
{
    // Build the table description.
    TableDesc td ("", "1", TableDesc::Scratch);
    td.addColumn (ArrayColumnDesc<float>  ("Pol", 1));
    td.addColumn (ArrayColumnDesc<float>  ("Freq", 1));
    td.addColumn (ArrayColumnDesc<float>  ("Data", 2));
    td.addColumn (ArrayColumnDesc<float>  ("Weight", IPosition(2,16,25),
					   ColumnDesc::FixedShape));
    td.defineHypercolumn ("TSMExample",
			  2,
			  stringToVector ("Data,Weight"),
			  stringToVector ("Pol,Freq"));
    
    // Now create a new table from the description.
    SetupNewTable newtab("tTiledCellStMan_tmp.data", td, Table::New);
    // Create a storage manager for it.
    TiledCellStMan sm1 ("TSMExample", IPosition(2,5,6));
    newtab.bindAll (sm1);
    Table table(newtab, 0, False, Table::LocalEndian, tsmOpt);

    Vector<float> freqValues(25);
    Vector<float> polValues(16);
    indgen (freqValues, float(200));
    indgen (polValues, float(300));
    ArrayColumn<float> freq (table, "Freq");
    ArrayColumn<float> pol (table, "Pol");
    ArrayColumn<float> data (table, "Data");
    ArrayColumn<float> weight (table, "Weight");
    Matrix<float> array(IPosition(2,16,25));
    Matrix<float> result(IPosition(2,16,25));
    uInt i;
    indgen (array);
    for (i=0; i<5; i++) {
	table.addRow();
	cout << " pol.isDefined=" << pol.isDefined(i) << endl;
	pol.setShape (i, IPosition(1,16), IPosition(1,1));
	cout << " pol.isDefined=" << pol.isDefined(i) << endl;
	cout << "data.isDefined=" << data.isDefined(i) << endl;
	data.setShape (i, IPosition(2,16,25), IPosition(2,12,10));
	cout << "weig.isDefined=" << weight.isDefined(i) << endl;
	cout << "freq.isDefined=" << freq.isDefined(i) << endl;
	cout << pol.shape(i) << freq.shape(i) << data.shape(i)
	     << weight.shape(i) << endl;
	data.put (i, array);
	weight.put (i, array+float(100));
	freq.put (i, freqValues);
	pol.put (i, polValues);
	array += float(200);
	freqValues += float(200);
	polValues += float(200);
    }
}

void writeNoHyper(const TSMOption& tsmOpt)
{
    // Build the table description.
    TableDesc td ("", "1", TableDesc::Scratch);
    td.addColumn (ArrayColumnDesc<float>  ("Pol", IPosition(1,16),
					   ColumnDesc::FixedShape));
    td.addColumn (ArrayColumnDesc<float>  ("Freq", 1, ColumnDesc::FixedShape));
    td.addColumn (ArrayColumnDesc<float>  ("Data", 2, ColumnDesc::FixedShape));
    td.addColumn (ArrayColumnDesc<float>  ("Weight", IPosition(2,16,25),
					   ColumnDesc::FixedShape));
    
    // Now create a new table from the description.
    SetupNewTable newtab("tTiledCellStMan_tmp.data", td, Table::New);
    // Create a storage manager for it.
    TiledCellStMan sm1 ("TSMExample", IPosition(2,5,6));
    newtab.setShapeColumn ("Freq", IPosition(1,25));
    newtab.setShapeColumn ("Data", IPosition(2,16,25));
    newtab.bindColumn ("Data", sm1);
    newtab.bindColumn ("Weight", sm1);
    Table table(newtab, 0, False, Table::AipsrcEndian, tsmOpt);

    Vector<float> freqValues(25);
    Vector<float> polValues(16);
    indgen (freqValues, float(200));
    indgen (polValues, float(300));
    ArrayColumn<float> freq (table, "Freq");
    ArrayColumn<float> pol (table, "Pol");
    ArrayColumn<float> data (table, "Data");
    ArrayColumn<float> weight (table, "Weight");
    Matrix<float> array(IPosition(2,16,25));
    Matrix<float> result(IPosition(2,16,25));
    uInt i;
    indgen (array);
    for (i=0; i<101; i++) {
	table.addRow();
	data.put (i, array);
	weight.put (i, array+float(100));
	freq.put (i, freqValues);
	pol.put (i, polValues);
	array += float(200);
	freqValues += float(200);
	polValues += float(200);
    }
    indgen (array);
    indgen (freqValues, float(200));
    indgen (polValues, float(300));
    for (i=0; i<table.nrow(); i++) {
	data.get (i, result);
	if (! allEQ (array, result)) {
	    cout << "mismatch in data row " << i << endl;
	}
	weight.get (i, result);
	if (! allEQ (array + float(100), result)) {
	    cout << "mismatch in weight row " << i << endl;
	}
	if (! allEQ (freq(i), freqValues)) {
	    cout << "mismatch in freq row " << i << endl;
	}
	if (! allEQ (pol(i), polValues)) {
	    cout << "mismatch in pol row " << i << endl;
	}
	array += float(200);
	freqValues += float(200);
	polValues += float(200);
    }
}
