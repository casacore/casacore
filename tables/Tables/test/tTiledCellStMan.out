 pol.isDefined=0
 pol.isDefined=1
data.isDefined=0
weig.isDefined=1
freq.isDefined=1
[16][25][16, 25][16, 25]
 pol.isDefined=0
 pol.isDefined=1
data.isDefined=0
weig.isDefined=1
freq.isDefined=1
[16][25][16, 25][16, 25]
 pol.isDefined=0
 pol.isDefined=1
data.isDefined=0
weig.isDefined=1
freq.isDefined=1
[16][25][16, 25][16, 25]
 pol.isDefined=0
 pol.isDefined=1
data.isDefined=0
weig.isDefined=1
freq.isDefined=1
[16][25][16, 25][16, 25]
 pol.isDefined=0
 pol.isDefined=1
data.isDefined=0
weig.isDefined=1
freq.isDefined=1
[16][25][16, 25][16, 25]
 pol.isDefined=1
 pol.isDefined=1
data.isDefined=1
weig.isDefined=1
freq.isDefined=1
[16][25][16, 25][16, 25]
 pol.isDefined=1
 pol.isDefined=1
data.isDefined=1
weig.isDefined=1
freq.isDefined=1
[16][25][16, 25][16, 25]
 pol.isDefined=1
 pol.isDefined=1
data.isDefined=1
weig.isDefined=1
freq.isDefined=1
[16][25][16, 25][16, 25]
 pol.isDefined=1
 pol.isDefined=1
data.isDefined=1
weig.isDefined=1
freq.isDefined=1
[16][25][16, 25][16, 25]
 pol.isDefined=1
 pol.isDefined=1
data.isDefined=1
weig.isDefined=1
freq.isDefined=1
[16][25][16, 25][16, 25]
